examples/chip-tool/config/PersistentStorage.cpp

```diff
@@ -204,3 +208,44 @@ LogCategory PersistentStorage::GetLoggingLevel()
 
     return chipLogLevel;
 }
+
+NodeId PersistentStorage::GetNodeId(const char * key, NodeId defaultVal)
```

Comment:
Why not compute this from the NOC every time, rather than store one in persistent storage (which works maybe for CHIP-tool in single fabric, but is not a very good example for the general case)?

Reply:
From which NOC?

The way chip-tool works is that you run one of the "pairing" commands, it commissions a device, remembers that device, and then when you run one of the other commands it talks to that device.  In particular, cluster commands do not provide a node id on the command line.  Before my changes this works by always using the same node id to talk to, and commissioning the node to have that fixed id.  With my changes we commission to a random id (good), but then we have to remember what id we used so the next command uses that same id to talk to.

Reply:
Noted.  Still, the fact the commissioner can call into the OpCreds delegate with this is fundamentally flawed:

```
    virtual CHIP_ERROR GenerateNodeOperationalCertificate(const Optional<NodeId> & nodeId, FabricId fabricId, const ByteSpan & csr, const ByteSpan & DAC, Callback::Callback<NOCGenerated> * onNOCGenerated) = 0;
```

It's not for the sdk to propose node IDs to PKI infrastructure.  This happens to work here because the example OpCreds delegate will accept node IDs and embed these into cert DNs.  However, real PKI infrastructure might not (the infrastructure I deal with does not).

This means that this approach is likely incompatible with a variant of the chip-tool that calls into an ecosystem-specific OpCreds delegate.  Such a thing is easy to build and useful (I have done this).

This problem could be solved by instead storing the node ID that comes back in the signed certs, whatever that may be.  I think this gets to the spirit of Tennessee's comment.

Reply:
Yes, in a world in which the "get me a cert" API had node id as output, not input, that would make sense.  That's not the API we have right now, as you note, but once we have that API shape I agree this code should extract the node id from the NOC it gets back.